lib/Sema/CSSimplify.cpp

```diff
@@ -2080,13 +2080,16 @@ ConstraintSystem::matchFunctionResultTypes(Type expectedResult, Type fnResult,
     auto anchor = locator.getAnchor();
     if (auto *callExpr = getAsExpr<CallExpr>(anchor)) {
       if (auto *innerCall = getAsExpr<CallExpr>(callExpr->getSemanticFn())) {
-        auto *innerCalleeLoc =
+        if (!simplifyType(getType(innerCall->getFn()))
+                 ->is<TypeVariableType>()) {
```

Comment:
I think this is wrong for the cases with IUO because this means that the follow-up logic that checks selected overload won't have it and the solver won't potentially end up not adding disjunction for the IUO. We need a test-case that verifies that...

I think this logic needs to be moved to `simplifyApplicableFnConstraint` and if the inner call doesn't have an overload selected yet we need to delay simplication of `ApplicableFunction` constraint until it is selected.